### PR TITLE
Add Build Guidances

### PR DESCRIPTION
**Pull request by wtam2018**
### What does this PR do?
This PR adds `BuildGuidances` which includes `Dockerfile` and `SourceToImage` for build strategies.   It is added to top level so that they can be shared by multiple projects.

### What issues does this PR fix or reference?
#51 

### Is your PR tested? Consider putting some instruction how to test your changes
Not applicable.

### CODE REVIEW

File: pkg/apis/workspaces/v1alpha2/buildGuidances.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+	// +optional
+	// Allows specifying a builder image to initiate s2i (SourceToImage) build
+	SourceToImage *SourceToImage `json:"s2i,omitempty"`
+}
+
+type Dockerfile struct {
+
+	// Name that allows referencing a build guidance
+	Name string `json:"name"`
```
Comment (sleshchenko) [inline pkg/apis/workspaces/v1alpha2/buildGuidances.go:19]: In terms of overriding does it makes sense move the name to the upper level, like we do with components:
```
components:
  - name: maven
    container: ....
```
? I'm not sure about just a question

Comment (davidfestal) [reply pkg/apis/workspaces/v1alpha2/buildGuidances.go:19]: The name is mandatory at the upper-level to identity elements in a list. When the field is the patchMergeKey of the list.
Here there is only 1 build guidance in a project afaik.

However I see you defined a union @wtam2018 Are we sure that the 2 values (dockerfile or s2i) are really mutually-exclusive ?

Comment (wtam2018) [reply pkg/apis/workspaces/v1alpha2/buildGuidances.go:19]: @davidfestal I'll have union removed dockerfile and s2i are not mutually excluded.   I should probably rename guidance to its plural form.

Comment (wtam2018) [reply pkg/apis/workspaces/v1alpha2/buildGuidances.go:19]: @davidfestal @sleshchenko Based on the feedbacks from our previous meeting, build guidance should be similar and at the same level as component/command.  I will rework this and take build guidance out of project.

File: pkg/apis/workspaces/v1alpha2/buildGuidances.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	SourceToImage *SourceToImage `json:"s2i,omitempty"`
+}
+
+type Dockerfile struct {
+	// Dockerfile location which can be an URL or a path relative to buildContext
+	DockerfileLocation string `json:"dockerfileLocation"`
+
+	// +optional
+	// Path of source directory to establish build context.  Default to the top level directory.
+	BuildContext string `json:"buildContext,omitempty"`
```
Comment (maysunfaisal) [inline pkg/apis/workspaces/v1alpha2/buildGuidances.go:44]: I am not sure if we want to continue to work on this branch for Build Guidance or open a new PR but for record, I had a question:

What is the difference between this `BuildContext` and the one provided from the Console POC import from devfile form https://github.com/openshift/console/pull/6321

Screen Shot 2020-11-16 at 12 29 19 PM

**Pull request closed without merging**